openedx/adg/lms/applications/tests/test_analytics.py

```diff
@@ -0,0 +1,277 @@
+    """
+    Test that the course started event is generated for course enrollment
+    """
+    mock_testing_env = mocker.patch('common.djangoapps.student.models.is_testing_environment')
+    mock_course_started_emit = mocker.patch('common.djangoapps.student.models.emit_course_progress_event')
+    mock_testing_env.return_value = False
+
+    course_to_enroll = courses['test_course1']
+    CourseEnrollment.enroll(user, course_to_enroll.id)
+    mock_course_started_emit.assert_called_once_with(user, course_to_enroll)
+
+
+@pytest.mark.parametrize('is_program_prereq', [True, False])
+@pytest.mark.django_db
+def test_program_prereq_course_progress_emit_for_enrollment(mocker, user, is_program_prereq):
+    """
+    Test that an emit is generated when a program prereq course is enrolled
+    """
+    mock_course_language = mocker.patch('openedx.adg.lms.applications.analytics.get_language_info')
+    mock_course_language.return_value = {'name': 'English'}
+    program_prereq_emit = mocker.patch(
+        'openedx.adg.lms.applications.analytics.emit_enrollment_event_for_program_prereq_courses'
+    )
+    course_group = MultilingualCourseGroupFactory(is_program_prerequisite=is_program_prereq)
+    course_to_enroll = MultilingualCourseFactory(multilingual_course_group=course_group).course
+
+    emit_course_progress_event(user, course_to_enroll)
+    if is_program_prereq:
+        program_prereq_emit.assert_called_once_with(user.id, course_to_enroll, COURSE_STARTED)
+    else:
+        program_prereq_emit.assert_not_called()
+
+
+def add_course_grade_for_user(user, course, is_passed=False):
+    """
+    Create a course enrollment and compute course grade for the given course
+    and user pair. Grade is passed or failed based on the `is_passed` value
+
+    Arguments:
+        user (User): User to compute the grade for
+        course (CourseOverview): Course to compute the grade for
+        is_passed (Boolean): Should the user pass in the course or not
+
+    Returns:
+        PersistentCourseGrade: Computed grade based on the given information
+    """
+    CourseEnrollmentFactory(user=user, course=course)
+    kwargs = {
+        'passed': is_passed,
+        'percent_grade': 100 if is_passed else 0,
+        'letter_grade': 'Pass' if is_passed else ''
+    }
+    grade_achieved = PersistentCourseGrade.update_or_create(user.id, course.id, **kwargs)
+    return grade_achieved
+
+
+@pytest.mark.parametrize('is_passed', [True, False])
+@pytest.mark.django_db
+def test_course_completion_emit_when_user_passes_or_fails(is_passed, user, courses, mocker):
+    """
+    Test that the course completion event is emitted when a user completed a course and
+    grade is computed
+    """
+    mock_testing_env = mocker.patch('lms.djangoapps.grades.models.is_testing_environment')
+    mock_testing_env.return_value = False
+    mock_course_completed_emit = mocker.patch('lms.djangoapps.grades.models.emit_course_progress_event')
```

Comment:
it won't always be a course completed emit, not in case course is failed?

I think we can rename it to `mock_course_progress_emit` or something else?

Reply:
Course completed means either passing or failing the course, so it is called `mock_course_completed_emit` in that sense